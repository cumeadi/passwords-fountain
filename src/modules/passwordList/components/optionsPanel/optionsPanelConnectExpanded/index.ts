export { OptionsPanelConnectExpanded } from './optionsPanelConnectExpanded.component';
