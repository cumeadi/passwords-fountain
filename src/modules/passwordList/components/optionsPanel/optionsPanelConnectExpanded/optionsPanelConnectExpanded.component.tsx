import { h, VNode } from 'preact';
import { TypedComponent } from '@/common/typings/prop-types';
import PropTypes from 'prop-types';
import { Wrapper } from './optionsPanelConnectExpanded.styles';
import { ButtonWrapper, Content, ContentWrapper } from '../optionsPanel.styles';
import { Button } from '@/common/components/button';
import { Text } from '@/modules/localisation/components/text';
import { VariantProps } from '../optionsPanel.component';
import { variantNames } from '@/modules/passwordList/passwordList.contants';
import { FormControl } from '@/common/components/formControl';
import { TextInput } from '@/common/components/textInput';
import { useRef, useState } from 'preact/hooks';
import { validateInputField } from '@/common/utils/form';

const formValidation = {
    masterKey(val?: string) {
        return (val && val.length >= 6) || 'optionsPanel.masterKeyTooShort';
    },
} as const;

export const OptionsPanelConnectExpanded: TypedComponent<VariantProps> = ({
    switchCurrentVariantName,
}: VariantProps) => {
    const formRef = useRef(undefined as any);
    const [masterKeyValue, setMasterKeyValue] = useState('');
    const [masterKeyErrors, setMasterKeyErrors] = useState('');

    const handleCancelClick = (): void =>
        switchCurrentVariantName(variantNames.connectCollapsed);

    const handleConfirmClick = (): void => {
        // TODO: dispatchAction decode
        // TODO: useSelector to check if decoded successfully or not
        const successfulDecode = true;
        if (successfulDecode) {
            switchCurrentVariantName(variantNames.entityFormCollapsed);
        } else {
            // TODO: Show error snackbar
        }
    };

    const renderInput = (): VNode => (
        <TextInput
            placeholder="e.g. MyStrongPassword1234"
            hasError={Boolean(masterKeyErrors)}
            name="masterKey"
            value={masterKeyValue}
            onInput={validateInputField(
                'masterKey',
                formRef,
                formValidation,
                setMasterKeyValue,
                setMasterKeyErrors
            )}
        />
    );

    const renderLabel = (): VNode => <Text>optionsPanel.enterMasterKey</Text>;
    const renderError = (): VNode => <Text>{masterKeyErrors}</Text>;
    return (
        <Wrapper>
            <ContentWrapper>
                <Content>
                    <form ref={formRef}>
                        <FormControl
                            hasError={Boolean(masterKeyErrors)}
                            renderLabel={renderLabel}
                            renderInput={renderInput}
                            renderError={renderError}
                        />
                    </form>
                </Content>
            </ContentWrapper>
            <ButtonWrapper>
                <Button onClick={handleCancelClick}>
                    <Text>optionsPanel.cancel</Text>
                </Button>
                <Button
                    onClick={handleConfirmClick}
                    disabled={Boolean(masterKeyErrors)}
                >
                    <Text>optionsPanel.confirm</Text>
                </Button>
            </ButtonWrapper>
        </Wrapper>
    );
};

OptionsPanelConnectExpanded.propTypes = {
    switchCurrentVariantName: PropTypes.func.isRequired,
};
