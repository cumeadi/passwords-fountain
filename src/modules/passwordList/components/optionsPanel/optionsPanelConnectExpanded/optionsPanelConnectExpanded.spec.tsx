import '@testing-library/jest-dom/extend-expect';
import { h } from 'preact';
import { render } from '@testing-library/preact';
import { OptionsPanelConnectExpanded } from './optionsPanelConnectExpanded.component';
import { StoreProvider } from '@preact-hooks/unistore';
import createStore from 'unistore';

const stateMock = {};

describe('OptionsPanelConnectExpanded', () => {
    it('should render correctly', () => {
        const { asFragment } = render(
            <StoreProvider value={createStore(stateMock)}>
                <OptionsPanelConnectExpanded
                    switchCurrentVariantName={(): void => {
                        /* eslint-disable @typescript-eslint/no-empty-function */
                    }}
                />
            </StoreProvider>
        );

        expect(asFragment()).toMatchSnapshot();
    });
});
