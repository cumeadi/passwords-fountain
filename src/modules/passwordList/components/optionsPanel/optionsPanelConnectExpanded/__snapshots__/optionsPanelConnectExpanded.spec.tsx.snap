// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OptionsPanelConnectExpanded should render correctly 1`] = `
<DocumentFragment>
  <section
    class="go115679"
  >
    <div
      class="go115679"
    >
      <div
        class="go4038517794"
      >
        <form>
          <section
            class="go3154290440"
          >
            <label
              class="go720639114"
              for="masterKey"
            >
              Enter your master key:
            </label>
            <div
              class="go3253370986"
            >
              <input
                aria-describedby="masterKey-error-message"
                aria-invalid="false"
                autofocus=""
                class="go3854179087"
                id="masterKey"
                name="masterKey"
                placeholder="e.g. MyStrongPassword1234"
                type="password"
              />
            </div>
          </section>
        </form>
      </div>
    </div>
    <div
      class="go3306721962"
    >
      <button
        class="go3135434547"
        type="button"
      >
        <span
          class="go4062099715"
        >
          Cancel
        </span>
      </button>
      <button
        class="go3135434547"
        disabled=""
        type="button"
      >
        <span
          class="go4062099715"
        >
          Confirm
        </span>
      </button>
    </div>
  </section>
</DocumentFragment>
`;
